src/value.jl

```diff
@@ -1,145 +1,181 @@
+iscatvalue(::Type{<:CatValue}) = IsCatValue
+valtype(::Type) = Union{} # no value type if not a proper categorical value type
+valtype(::Type{<:CategoricalValue{T}}) where {T} = T
+valtype(::Type{<:CategoricalString}) = String
+
+# integer type of category reference codes for given type
+reftype(::Type) = DefaultRefType
+# TODO reftype(::Type{T}) where {T <: Integer} = T ?
+reftype(::Type{<:CatValue{R}}) where {R} = R
+
+pool(x::CatValue) = x.pool
+level(x::CatValue) = x.level
+
+# extract the type of the original value from array eltype `T`
+unwrap_catvalue_type(::Type{T}) where {T} =
+    unwrap_catvalue_type(iscatvalue(T), T)
+
+function unwrap_catvalue_type(::Type{T}) where T >: Null
+    V = Nulls.T(T)
+    Union{unwrap_catvalue_type(iscatvalue(V), V), Null}
 end
 
-Base.convert(::Type{CategoricalValue{T, R}}, x::CategoricalValue{T, R}) where {T, R <: Integer} = x
-Base.convert(::Type{CategoricalValue{T}}, x::CategoricalValue{T}) where {T} = x
-Base.convert(::Type{CategoricalValue}, x::CategoricalValue) = x
+unwrap_catvalue_type(::Type{Any}) = # to prevent dispatching to T>:Null method
+    unwrap_catvalue_type(NotCatValue, Any)
+unwrap_catvalue_type(::Type{NotCatValue}, ::Type{T}) where {T} = T
+unwrap_catvalue_type(::Type{IsCatValue}, ::Type{T}) where {T} = valtype(T)
 
-Base.convert(::Type{Union{CategoricalValue{T, R}, Null}}, x::CategoricalValue{T, R}) where {T, R <: Integer} = x
-Base.convert(::Type{Union{CategoricalValue{T}, Null}}, x::CategoricalValue{T}) where {T} = x
-Base.convert(::Type{Union{CategoricalValue, Null}}, x::CategoricalValue) = x
+# get default categorical value type given value type `T` and reference type `R`
+function catvalue_type(::Type{T}, ::Type{R}) where {T, R}
+    V = Nulls.T(T)
+    catvalue_type(iscatvalue(V), V, R)
+end
+
+catvalue_type(::Type{IsCatValue}, ::Type{T}, ::Type{R}) where {T, R} =
+    reftype(T) === R ? T : catvalue_type(valtype(T), R)
+catvalue_type(::Type{NotCatValue}, ::Type{T}, ::Type{R}) where {T, R} =
+    CategoricalValue{T, R}
+catvalue_type(::Type{NotCatValue}, ::Type{<:AbstractString}, ::Type{R}) where {R} =
+    CategoricalString{R}
+
+Base.get(x::CatValue) = index(pool(x))[level(x)]
+order(x::CatValue) = order(pool(x))[level(x)]
+
+# creates categorical value for `level` from the `pool`
+# The result is of type `C` that has "categorical value" trait
+catvalue(level::Integer, pool::CategoricalPool{T, R, C}) where {T, R, C} =
+    C(convert(R, level), pool)
 
-Base.promote_rule(::Type{CategoricalValue{S, R}}, ::Type{T}) where {S, T, R} = promote_type(S, T)
-Base.promote_rule(::Type{CategoricalValue{S}}, ::Type{T}) where {S, T} = promote_type(S, T)
-Base.promote_rule(::Type{CategoricalValue}, ::Type{T}) where {T} = T
+Base.convert(::Type{T}, x::T) where {T <: CatValue} = x
+Base.convert(::Type{Union{T, Null}}, x::T) where {T <: CatValue} = x
+
+# FIXME do we need this rule or promotion is only required for CategoricalString?
```

Comment:
If there aren't any test failures, we can assume it's enough. I think the ambiguities come from the fact that `CategoricalString <: AbstractString`.

Reply:
This line is needed to pass tests in `05_convert.jl`, e.g.
```julia
pool = CategoricalPool([1, 2, 3])
v1 = catvalue(1, pool)
@test promote(1, v1) === (1, 1)
```
So if you want to preserve this behaviour for non-string factors, we should keep it.
The other test sets don't depend on it.

Reply:
Yes, I think we want to preserve this behavior. I thought your comment meant that tests passed without it.